Task: Spring Boot Portfolio Project — Part 10.Add Order model. Executor: Oleksii Bashkirov.

OleksiiBashkirov opened the pull request:
feat: Implement Order and OrderItem entities with soft delete and relationships

- Created Order and OrderItem entities
- Implemented relationships for Order and OrderItem entities
- Added soft delete functionality to Order entity
- Created OrderRepository and OrderItemRepository
- Implemented OrderService and OrderServiceImpl
- Added endpoints for managing orders and order items in OrderController
- Updated Liquibase changelog files:
  - Renamed changelog files for better understanding of their contents
  - Split categories and books_categories tables into separate changelog files

OleksiiBashkirov committed:
Fixed security configuration and liquibase issue

OleksiiBashkirov committed:
Fixed checkstyle errors

OleksiiBashkirov committed:
- Removed the redundant Repository annotation in the RoleRepository interface.

OleksiiBashkirov committed:
- Updated pom.xml file;
- Added annotations in SecurityConfig;
- Included userDetailsService in SecurityConfig;
- Removed EAGER fetch type for roles in User entity;
- Updated CustomUserDetailsService to use EntityNotFoundException;
- Implemented role assignment in UserServiceImpl during user registration.

OleksiiBashkirov committed:
- Removed debug messages.

OleksiiBashkirov committed:
1. Updated SecurityConfig.java:
   - Removed the @EnableWebSecurity annotation.

2. Updated BookMapper.java:
   - Removed the @Component annotation.

3. Updated User.java:
   - Corrected @ManyToMany annotation formatting.

4. Updated CustomUserDetailsService.java:
   - Refactored the "loadUserByUsername" method.

5. Updated UserServiceImpl.java:
   - Replaced Collections.singleton with Set.of for setting user roles.

OleksiiBashkirov committed:
feat: Implement JWT authentication

- Added `login(UserLoginRequestDto request)` method to `AuthenticationController`.
- Created `AuthenticationService`, `JwtAuthenticationFilter`, and `JwtUtil` classes.
- Made `/auth/login` endpoint publicly accessible.
- Configured `SecurityFilterChain` to use the new JWT filter.
- Implemented data validation for new DTO classes.

OleksiiBashkirov committed:
Implement suggestions and improvements

- Removed empty lines in pom.xml;
- Removed JwtConfig.java;
- Simplified SecurityConfig.java by adjusting the security filter chain and configurations;
- Added constants for "BearerAuth" and "Bearer " in appropriate classes;
- Updated UserLoginResponseDto;
- Enhanced UserRepository with @EntityGraph for roles in findByEmail method;
- Replaced hardcoded "Authorization" header string with HttpHeaders.AUTHORIZATION constant;
- Ensured proper substring usage in JwtAuthenticationFilter.

OleksiiBashkirov committed:
- Deleted unnecessary class `JwtConfig`;
- Refactored constant `BEARER_AUTH` in `SwaggerConfig`;

OleksiiBashkirov committed:
- Added constants for "bearer" and "JWT" in SwaggerConfig class;

OleksiiBashkirov committed:
- Fixed checkstyle errors;

OleksiiBashkirov committed:
feat: Added support for Category entity and related functionalities

- Added `Category` entity
- Updated `Book` class to include `private Set<Category> categories = new HashSet<>()`
- Added method to `BookRepository`: findAllByCategoryId

- Implemented `CategoryRepository` interface using `JpaRepository`
- Implemented DTO classes for `Category` entity
- Modified `BookMapper` class:
  - Added `BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book)`
  - Added `@AfterMapping default void setCategoryIds(@MappingTarget BookDto bookDto, Book book)`

- Added `CategoryMapper` class:
  - Added `CategoryDto toDto(Category category)`
  - Added `Category toModel(CategoryDto categoryDTO)`

- Implemented `CategoryService` interface and `CategoryServiceImpl` class:
  - Added `List<CategoryDto> findAll()`
  - Added `CategoryDto getById(Long id)`
  - Added `CategoryDto save(CategoryDto categoryDto)`
  - Added `CategoryDto update(Long id, CategoryDto categoryDto)`
  - Added `void deleteById(Long id)`

- Added `CategoryController` class:
  - Added `CategoryDto createCategory(CategoryDto categoryDto)`
  - Added `List<CategoryDto> getAll()`
  - Added `CategoryDto getCategoryById(Long id)`
  - Added `CategoryDto updateCategory(Long id, CreateCategoryRequestDto requestDto)`
  - Added `void deleteCategory(Long id)`
  - Added `Page<BookDtoWithoutCategoryIds> getBooksByCategoryId(Long id)`

- Added Pagination, Sorting, and Swagger to all controllers
- Used Liquibase for database versioning and implement soft delete approach for entities

OleksiiBashkirov committed:
- Modified AuthenticationController;

OleksiiBashkirov committed:
Merge branch 'main' into add-category-model

OleksiiBashkirov committed:
- version 1.0(2024/07/30 4:18PM)

OleksiiBashkirov committed:
fix: made necessary corrections

- Deleted UserLoginResponseDto class.
- Updated AuthenticationService interface to return String for login method.
- Modified AuthenticationServiceImpl to return JWT token directly as String.
- Adjusted AuthenticationController to return String for login method.
- Removed unnecessary changes in SecurityConfig.
- Corrected BookDtoWithoutCategoryIds and CategoryDto to remove isDeleted field.
- Updated Book model to use book_categories table for ManyToMany relationship with Category.
- Ensured Category model has unique constraint on name field.
- Fixed BookServiceImpl to use .toList() instead of .collect(Collectors.toList()).
- Fixed CategoryServiceImpl to use .toList() instead of .collect(Collectors.toList()).
- Added missing books_categories table to db.changelog-7.0.yaml for ManyToMany relationship between Book and Category.

OleksiiBashkirov committed:
- Merged with `add-category-model` branch to have the latest code version;

OleksiiBashkirov committed:
feat: Implement Shopping Cart functionality in the project

- Created ShoppingCart entity with appropriate fields and relationships:
  - Included fields id, user, cartItems, and isDeleted.
  - Configured one-to-one relationship with User entity.
  - Configured one-to-many relationship with CartItem entity.
  - Added soft delete functionality using @SQLDelete and @SQLRestriction annotations.

- Created CartItem entity to represent items in the shopping cart:
  - Included fields id, shoppingCart, book, and quantity.
  - Configured many-to-one relationships with ShoppingCart and Book entities.

- Updated User entity to include a one-to-one relationship with ShoppingCart:
  - Added shoppingCart field with @OneToOne and @JoinColumn annotations.

- Created ShoppingCartRepository interface extending JpaRepository for CRUD operations on ShoppingCart.

- Created CartItemRepository interface extending JpaRepository for CRUD operations on CartItem.

- Implemented ShoppingCartService interface with methods to manage the shopping cart:
  - getShoppingCartByUserId: Retrieve shopping cart by user ID, creating one if it doesn't exist.
  - addToCart: Add a book to the shopping cart.
  - updateCartItem: Update the quantity of a cart item.
  - removeCartItem: Remove a cart item from the shopping cart.

- Implemented ShoppingCartServiceImpl class to provide the actual implementation of ShoppingCartService methods.

- Created ShoppingCartController with endpoints to manage the shopping cart:
  - GET /api/cart: Retrieve user's shopping cart.
  - POST /api/cart: Add book to the shopping cart.
  - PUT /api/cart/cart-items/{cartItemId}: Update quantity of a book in the shopping cart.
  - DELETE /api/cart/cart-items/{cartItemId}: Remove a book from the shopping cart.

- Added validation DTOs:
  - AddToCartRequest for adding books to the cart.
  - UpdateCartItemRequest for updating cart item quantity.

- Updated Liquibase changelog files to create necessary tables and constraints:
  - db.changelog-8.0.yaml: Added creation of shopping_cart table.
  - db.changelog-9.0.yaml: Added creation of cart_item table.

- Ensured proper role-based access control for new endpoints in the security configuration:
  - Users with role USER can access cart-related endpoints.
  - Users with role ADMIN can manage books and categories.

OleksiiBashkirov committed:
Merge branch 'main' into add-shopping-cart-model

OleksiiBashkirov committed:
- Branch startpoint

OleksiiBashkirov committed:
feat: Implement Order and OrderItem entities with soft delete and relationships

- Created Order and OrderItem entities
- Implemented relationships for Order and OrderItem entities
- Added soft delete functionality to Order entity
- Created OrderRepository and OrderItemRepository
- Implemented OrderService and OrderServiceImpl
- Added endpoints for managing orders and order items in OrderController
- Updated Liquibase changelog files:
  - Renamed changelog files for better understanding of their contents
  - Split categories and books_categories tables into separate changelog files

OleksiiBashkirov committed:
Merge branch 'add-category-model' into add-shopping-cart-model

okuzan commented on src/main/java/mate/academy/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Override
+    public List<OrderItem> getOrderItems(Long orderId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("Order not found by id: "
+                                + orderId));
```
repreated logic , extract to a private method

okuzan commented on src/main/java/mate/academy/model/Order.java:
```diff
@@ -0,0 +1,67 @@
+@Table(name = "orders")
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id = ?")
+@SQLRestriction(value = "is_deleted = false")
```
```suggestion
@SQLDelete(sql = "UPDATE orders SET is_deleted = TRUE WHERE id = ?")
@SQLRestriction(value = "is_deleted = FALSE")
```

OleksiiBashkirov committed:
Title: Refactor Service implementations and Update Model annotations

- Extracted repeated logic into private methods in `CategoryServiceImpl` and `BookServiceImpl`.
- Corrected SQL annotations for consistency in the `Category`, `Book`, `Role`, and `User` models:
    * Changed `@SQLDelete` statements to use `TRUE` instead of `true`.
    * Updated `@SQLRestriction` to use `FALSE` instead of `false`.

OleksiiBashkirov committed:
Merge branch 'add-category-model' into add-shopping-cart-model

OleksiiBashkirov committed:
Refactored SecurityConfig and Other Classes

- Grouped endpoints into separate methods in the configureAuthorization method
- Refactored code based on feedback from previous tasks in this project

OleksiiBashkirov committed:
- Resolved merge conflicts in add-shopping-cart-model

OleksiiBashkirov committed:
Merge branch 'add-shopping-cart-model' into add-order-model-hw

OleksiiBashkirov committed:
feat: Refactor and improve code based on mentor feedback

- Extracted repeated logic into a private method in OrderServiceImpl.java
- Updated @SQLDelete and @SQLRestriction annotations in Order.java for consistency
- Refactored other classes for cleaner and more readable code by creating separate methods without altering existing logic

OleksiiBashkirov committed:
- Merged some improvements from the add-shopping-cart-model branch into add-order-model-hw

MentorMateAcademy commented on src/main/java/mate/academy/controller/ShoppingCartController.java:
```diff
@@ -27,16 +28,16 @@ public class ShoppingCartController {
     public ResponseEntity<Void> addToCart(@AuthenticationPrincipal UserPrincipal userPrincipal,
             @RequestBody AddToCartRequest request) {
-        shoppingCartService.addToCart(userPrincipal.getUser(), request.getBookId(),
+        shoppingCartService.addToCart(userPrincipal.getUser().getId(),request.getBookId(),
```
```suggestion
        shoppingCartService.addToCart(userPrincipal.getUser().getId(), request.getBookId(),
```

MentorMateAcademy replied:
formatting code

MentorMateAcademy commented on src/main/java/mate/academy/repository/BookRepository.java:
```diff
@@ -11,8 +11,7 @@
 import org.springframework.stereotype.Repository;
 
 @Repository
```
```suggestion
```

MentorMateAcademy replied:
this anno is redundant

MentorMateAcademy commented on src/main/java/mate/academy/repository/BookRepositoryCustomImpl.java:
```diff
@@ -18,18 +18,23 @@ public class BookRepositoryCustomImpl implements BookRepositoryCustom {
     public List<Book> findAllBySearchParams(Map<String, List<String>> params) {
-        CriteriaBuilder cb = em.getCriteriaBuilder();
-        CriteriaQuery<Book> cq = cb.createQuery(Book.class);
+        var cb = em.getCriteriaBuilder();
+        var cq = cb.createQuery(Book.class);
```
short naming is bad practice

MentorMateAcademy commented on src/main/java/mate/academy/service/impl/OrderServiceImpl.java:
```diff
@@ -0,0 +1,55 @@
+
+    @Override
+    public List<OrderItem> getOrderItems(Long orderId) {
+        var order = getOrder(orderId);
```
What profit of using var here?

MentorMateAcademy commented on src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -17,46 +16,46 @@
 
-    private ShoppingCart createNewShoppingCartForUser(Long userId) {
-        var user = findUserById(userId);
+    private ShoppingCart createShoppingCartForUser(Long userId) {
+        var user = userRepository.findById(userId).orElseThrow(
+                () -> new EntityNotFoundException("User not found by id: " + userId));
         var cart = new ShoppingCart();
         cart.setUser(user);
         return shoppingCartRepository.save(cart);
     }
```
private methods after public

MentorMateAcademy commented on src/main/java/mate/academy/service/impl/ShoppingCartServiceImpl.java:
```diff
@@ -17,46 +16,46 @@
         shoppingCartRepository.save(cart);
     }
 
+    private static CartItem createCartItem(int quantity, ShoppingCart cart, Book book) {
```
Why do you need static?

OleksiiBashkirov committed:
Made corrections:
- Removed redundant @Repository annotation from BookRepository.
- Improve variable naming in BookRepositoryCustomImpl.
- Use explicit types instead of var for clarity in OrderServiceImpl.
- Reorder methods in ShoppingCartServiceImpl: place private methods after public methods.
- Remove static modifier from createCartItem method in ShoppingCartServiceImpl.

OleksiiBashkirov committed:
Merge branch 'main' into add-order-model-hw

OleksiiBashkirov committed:
Update BookServiceImpl.java

- added missed curly parentheses
